### PR TITLE
Implemented Alerting API

### PR DESCRIPTION
**Pull request by lukaa12**
KNTK-176

### CODE REVIEW

File: kentikapi/internal/api_payloads/alerting.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	Response Response `json:"response"`
+}
+
+type Response struct {
+	Result string `json:"result"`
+}
+
+type GetActiveAlertsResponse []models.Alarm
+
+func (r GetActiveAlertsResponse) ToAlarms() ([]models.Alarm, error) {
```
Comment (mateuszmidor) [inline kentikapi/internal/api_payloads/alerting.go:35]: returning error probably not needed

File: kentikapi/models/convertible_bool.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package models
+
+type Bool bool
+
+func (bit *Bool) UnmarshalJSON(b []byte) error {
+	txt := string(b)
+	*bit = Bool(txt == "1" || txt == "true")
```
Comment (mateuszmidor) [inline kentikapi/models/convertible_bool.go:7]: you can also use strconv.ParseBool that covers a number of stringy bool values

**Review by mateuszmidor — APPROVED** [review]

File: kentikapi/models/convertible_bool.go (hunk 2 above)
Comment (Fenthick) [reply kentikapi/models/convertible_bool.go:7]: Please rebase onto master and use/modify BoolAsString

File: kentikapi/internal/api_payloads/alerting.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	return r, nil
+}
+
+type GetHistoricalAlertsResponse []models.HistoricalAlert
+
+func (r GetHistoricalAlertsResponse) ToHistoricalAlerts() ([]models.HistoricalAlert, error) {
```
Comment (Fenthick) [inline kentikapi/internal/api_payloads/alerting.go:41]: Error not needed.

File: kentikapi/internal/api_resources/alerting.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	if response.Response.Result != "OK" {
+		return errors.New("Creating Manual Mitigation Failed")
+	}
+
+	return nil
+}
+
+func (a *AlertingAPI) GetActiveAlerts(ctx context.Context, startTime time.Time, endTime time.Time, filterBy string, filterVal string, showMitigations bool,
```
Comment (Fenthick) [inline kentikapi/internal/api_resources/alerting.go:40]: Please group alerts parameters under AlertsParameters data structure and pass it to GetActiveAlerts and GetAlertsHistory

File: kentikapi/internal/api_resources/alerting.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	showAlarms bool, showMatches bool, learningMode bool) ([]models.Alarm, error) {
+	var response api_payloads.GetActiveAlertsResponse
+	timeFormatStr := "2006-01-02T15:04:05"
+	path := api_endpoints.GetActiveAlertsPath(startTime.Format(timeFormatStr), endTime.Format(timeFormatStr), filterBy, filterVal, boolToInt(showMitigations),
```
Comment (Fenthick) [inline kentikapi/internal/api_resources/alerting.go:44]: Please call `startTime.Format(timeFormatStr)` and `boolToInt()` inside  api_endpoints.GetActiveAlertsPath - this is implementation detail of formatting the query string.

File: kentikapi/internal/api_resources/alerting.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+func (a *AlertingAPI) CreateManualMitigation(ctx context.Context, mm models.ManualMitigation) error {
+	payload := api_payloads.ManualMitigationToPayload(mm)
+	var response api_payloads.CreateManualMitigationResponse
+
+	if err := a.PostAndValidate(ctx, api_endpoints.ManualMitigationPath, payload, &response); err != nil {
+		return err
+	}
+
+	if response.Response.Result != "OK" {
+		return errors.New("Creating Manual Mitigation Failed")
```
Comment (Fenthick) [inline kentikapi/internal/api_resources/alerting.go:34]: > Error strings should not be capitalized (unless beginning with proper nouns or acronyms) 

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: kentikapi/internal/api_resources/alerting_test.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+var time1 = time.Date(2020, time.January, 19, 13, 50, 0, 0, time.Local)
+var time2 = time.Date(2021, time.March, 19, 13, 50, 0, 0, time.Local)
+
+func TestCrerateManualMitigation(t *testing.T) {
+	createResponsePayload := `
+	{
+        "response": {
+            "result": "OK"
+        }
+    }`
```
Comment (Fenthick) [inline kentikapi/internal/api_resources/alerting_test.go:23]: Nit: formatting

File: kentikapi/internal/api_resources/alerting_test.go [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+	transport := &api_connection.StubTransport{ResponseBody: createResponsePayload}
+	alertingAPI := api_resources.NewAlertingAPI(transport)
+	mm := models.ManualMitigation{
+		IPCidr:                "192.168.0.0/24",
+		PlatformID:            1234,
+		MethodID:              12345,
+		Comment:               nil,
+		MinutesBeforeAutoStop: "20",
+	}
+
```
Comment (Fenthick) [inline kentikapi/internal/api_resources/alerting_test.go:33]: Please verify the request path and query. In other tests too.

Comment (Fenthick) [inline kentikapi/internal/api_resources/alerting_test.go:33]: Please verify the request body

File: kentikapi/internal/api_resources/alerting_test.go [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            "mit_alert_ip": "",
+            "mit_threshold_id": 0,
+            "mit_method_id": 0,
+            "args": "",
+            "id": 0,
+            "policy_id": 15094,
+            "policy_name": "test_policy1",
+            "alert_key_lookup": "443"
+        }
+    ]`
```
Comment (Fenthick) [inline kentikapi/internal/api_resources/alerting_test.go:74]: Nit: formatting

File: kentikapi/models/alerting.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+	AlertValue2nd   float32   `json:"alert_value2nd"`
+	AlertValue3rd   float32   `json:"alert_value3rd"`
+	AlertMatchCount int       `json:"alert_match_count"`
+	AlertBaseline   int       `json:"alert_baseline"`
+	AlertSeverity   string    `json:"alert_severity"`
+	BaselineUsed    int       `json:"baseline_used"`
+	LearningMode    Bool      `json:"learning_mode"`
+	DebugMode       Bool      `json:"debug_mode"`
+	AlarmStart      time.Time `json:"alarm_start"`
+	AlarmEnd        *string   `json:"alarm_end"`
```
Comment (Fenthick) [inline kentikapi/models/alerting.go:33]: Add omitempty for pointers

**Review by Fenthick — CHANGES_REQUESTED** [review]

**Review by mateuszmidor — APPROVED** [review]

File: kentikapi/internal/api_payloads/types.go [hunk 11]
```diff
@@ -32,27 +32,36 @@ func (p *IntAsString) UnmarshalJSON(data []byte) (err error) {
 	return nil
 }
 
-// boolAsString evens out deserialization of numbers represented in JSON document sometimes as bool and sometimes as string.
-type boolAsString bool
+// BoolAsStringOrInt evens out deserialization of numbers represented in JSON document sometimes as bool and sometimes as string.
```
Comment (Fenthick) [inline kentikapi/internal/api_payloads/types.go:35]:
```suggestion
// BoolAsStringOrInt evens out deserialization of numbers represented in JSON document as bool, string or number.
```

File: kentikapi/internal/api_payloads/types.go [hunk 12]
```diff
@@ -32,27 +32,36 @@ func (p *IntAsString) UnmarshalJSON(data []byte) (err error) {
...
+		if asString == "1" {
+			*p = true
+		} else if asString == "0" {
+			*p = false
+		} else {
+			return fmt.Errorf("BoolAsStringOrInt.UnmarshalJSON: parse bool unexpected value %v", value)
+		}
 	default:
-		return fmt.Errorf("boolAsString.UnmarshalJSON input should be bool or string, got %v (%T)", value, value)
+		return fmt.Errorf("BoolAsStringOrInt.UnmarshalJSON input should be bool or string, got %v (%T)", value, value)
```
Comment (Fenthick) [inline kentikapi/internal/api_payloads/types.go:64]:
```suggestion
		return fmt.Errorf("BoolAsStringOrInt.UnmarshalJSON input should be bool, string or number, got %v (%T)", value, value)
```

**Review by Fenthick — CHANGES_REQUESTED** [review]

File: kentikapi/internal/api_endpoints/alerting.go [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+import "fmt"
+
+const (
+	ManualMitigationPath = "/alerts/manual-mitigate"
+	AlertsActivePath     = "/alerts-active/alarms"
+	AlertsHistoryPath    = "/alerts-active/alerts-history"
+)
+
+func GetActiveAlertsPath(startTime string, endTime string, filterBy string, filterVal string, showMitigations int, showAlarms int, showMatches int, learningMode int) string {
+	return fmt.Sprintf("%v?startTime=%v&endTime=%v&filterBy=%v&filterVal=%v&showMitigations=%v&showAlarms=%v&showMatches=%v&learningMode=%v",
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/api_endpoints/alerting.go]: Done

File: kentikapi/internal/api_payloads/alerting.go [hunk 14]
```diff
@@ -0,0 +1,43 @@
+package api_payloads
+
+import (
+	"github.com/kentik/community_sdk_golang/kentikapi/models"
+)
+
+type CreateManualMitigationRequest struct {
+	IPCidr                string    `json:"ipCidr"`
+	Comment               *string   `json:"comment"`
```
[the remark this answers is not in the document]
Comment (Fenthick) [reply kentikapi/internal/api_payloads/alerting.go]: Done

File: kentikapi/internal/api_payloads/alerting.go (hunk 3 above)
Comment (Fenthick) [reply kentikapi/internal/api_payloads/alerting.go]: Done

File: kentikapi/internal/api_resources/alerting.go (hunk 4 above)
Comment (Fenthick) [reply kentikapi/internal/api_resources/alerting.go]: Done

File: kentikapi/internal/api_resources/alerting.go (hunk 5 above)
Comment (Fenthick) [reply kentikapi/internal/api_resources/alerting.go]: Done

File: kentikapi/internal/api_resources/alerting.go (hunk 6 above)
Comment (Fenthick) [reply kentikapi/internal/api_resources/alerting.go]: Done

File: kentikapi/internal/api_resources/alerting_test.go (hunk 8 above)
Comment (Fenthick) [reply kentikapi/internal/api_resources/alerting_test.go:35]: Done

Comment (Fenthick) [reply kentikapi/internal/api_resources/alerting_test.go:35]: Done

File: kentikapi/models/alerting.go (hunk 10 above)
Comment (Fenthick) [reply kentikapi/models/alerting.go]: Done

**Review by Fenthick — APPROVED** [review]
Apart from updating a docstring and error message it's good to go.

File: kentikapi/internal/api_payloads/types.go (hunk 11 above)
Comment (Fenthick) [reply kentikapi/internal/api_payloads/types.go]: Done

File: kentikapi/internal/api_payloads/types.go (hunk 12 above)
Comment (Fenthick) [reply kentikapi/internal/api_payloads/types.go]: Done

**Review by Fenthick — APPROVED** [review]

**Review by mateuszmidor — APPROVED** [review]

**Review by TedatKentik — APPROVED** [review]
alerting implementation

**Pull request merged**